### PR TITLE
Modify the ResourceGroupBasedService to return ArmrestCollection objects.

### PR DESCRIPTION
**Pull request by djberg96**
This is a solution to https://github.com/ManageIQ/azure-armrest/issues/178.

This modifies the ResourceGroupBasedService so that the `list` and `list_all` methods return an `ArmrestCollection` object instead of a plain array. In turn, the `ArmrestCollection` object has been modified to include a `response_headers` accessor, and a `ResponseHeaders` model has been added as well, which is then set within the `list` and `list_all` methods.

*Update: the ResponseHeaders class was already added within a separate PR, so we use that now. Also, the method was changed from :headers to :response_headers.*

The short version is that now you can do something like this:

```
results = vms.list_all
results.response_headers.x_ms_ratelimit_remaining_subscription_reads
```

This lets us store the returned header information per collection set, rather than per object, since such information would be redundant.

*Update: I should amend that to say that we want the header information set once per request. So, a :list or :list_all call would set it once for the collection, while a :get would set it once on the returned object.*

~~A couple specs are currently failing because of missing attributes, but they're minor.~~

### CODE REVIEW

**Comment by djberg96** [comment]
@bzwei What do you think?

File: lib/azure/armrest/armrest_collection.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@ module Azure
   module Armrest
     class ArmrestCollection < Array
       attr_accessor :continuation_token
+      attr_accessor :response_headers
```
Comment (bzwei) [inline lib/azure/armrest/armrest_collection.rb]: Add an initializer to accept the response_headers as an optional hash, and covert to model object internally. Then the calling can be greatly simplified.

File: lib/azure/armrest/model/base_model.rb [hunk 2]
```diff
@@ -22,6 +22,8 @@ def self.attr_hash(*attrs)
 
       attr_hash :tags
 
+      attr_accessor :response_headers
```
Comment (bzwei) [inline lib/azure/armrest/model/base_model.rb]: Similarly add response_headers as an optional hash argument in the initializer.
How useful are the headers? The model object if returned by a single GET has headers, but if in a list does not have headers. Will the difference cause trouble for application?

File: lib/azure/armrest/resource_group_based_service.rb [hunk 3]
```diff
@@ -2,25 +2,56 @@ module Azure
...
+      # Returns an ArmrestCollection, with the response headers set
+      # for the operation as a whole.
+      #
       def list(rgroup = configuration.resource_group)
         validate_resource_group(rgroup)
 
         url = build_url(rgroup)
         url = yield(url) || url if block_given?
+
         response = rest_get(url)
```
Comment (bzwei) [inline lib/azure/armrest/resource_group_based_service.rb]: I would expect all LIST methods from Azure API to have `continuation_token`. Otherwise how do they manage a super long list without paging?

File: lib/azure/armrest/resource_group_based_service.rb [hunk 4]
```diff
@@ -35,19 +66,34 @@ def list(rgroup = configuration.resource_group)
...
-        results = JSON.parse(response)['value'].map { |hash| model_class.new(hash) }
+        json_response = JSON.parse(response)['value']
+
+        results = ArmrestCollection.new(json_response.map { |hash| model_class.new(hash) })
+
         filter.empty? ? results : results.select { |obj| filter.all? { |k, v| obj.public_send(k) == v } }
+
+        results.response_headers = Azure::Armrest::ResponseHeaders.new(response.headers)
```
Comment (bzwei) [inline lib/azure/armrest/resource_group_based_service.rb]: You should use new `results` from above line. But it is no longer needed if you refactor the initializer as suggested.

File: lib/azure/armrest/model/base_model.rb (hunk 2 above)
Comment (djberg96) [inline lib/azure/armrest/model/base_model.rb]: We don't want to add headers for every object, because they would all be identical, and would create unnecessary memory bloat. It's meant to only be set once per request.

File: lib/azure/armrest/armrest_collection.rb (hunk 1 above)
Comment (djberg96) [inline lib/azure/armrest/armrest_collection.rb]: I generally avoid redefining initialize on a subclass of a core class because it can lead to unexpected behavior. Since this is only used internally, I don't see the advantage of redefining initialize.

File: lib/azure/armrest/resource_group_based_service.rb (hunk 3 above)
Comment (djberg96) [inline lib/azure/armrest/resource_group_based_service.rb]: I'm actually not entirely sure that Azure sets this in a uniform manner, or for all resources. Consider the way tokens are parsed for insights versus storage.

Also, I still have this issue outstanding: https://github.com/ManageIQ/azure-armrest/issues/154

Anyway, I agree in principal, but think it should be a separate PR for now.

File: lib/azure/armrest/armrest_collection.rb (hunk 1 above)
Comment (djberg96) [inline lib/azure/armrest/armrest_collection.rb]: Upon further review, maybe this would be easier, since we have a bunch of custom list methods scattered throughout.

File: lib/azure/armrest/resource_group_based_service.rb (hunk 3 above)
Comment (djberg96) [inline lib/azure/armrest/resource_group_based_service.rb]: After scouring the docs a bit more, I think we can just use parse_skip_token from armrest_service.rb for most cases.

File: lib/azure/armrest/model/base_model.rb (hunk 2 above)
Comment (bzwei) [inline lib/azure/armrest/model/base_model.rb]: Memory will not be a concern if every model object holds a reference to the same header instance. 

I assume the header responding for a LIST request is different from the header responding for a GET request for a single object.  I therefore agree with you that it may be wrong to assign an unprocessed LIST header to each individual model object, unless the two headers are identical.

In addition, we should hide the response_header for internal use only because of the above discussed difference. Is there a use case that the user should know the header?

File: lib/azure/armrest/resource_group_based_service.rb (hunk 4 above)
Comment (djberg96) [inline lib/azure/armrest/resource_group_based_service.rb]: Agreed.

**Comment by djberg96** [comment]
@bzwei I've updated the ArmrestCollection class to accept a response object and a class. It then parses and sets the information internally, and returns an array of the class type provided.

Please let me know if this is what you had in mind. There are a few event specs that are failing now that I might need some help with.

File: lib/azure/armrest/insights/event_service.rb [hunk 5]
```diff
@@ -44,22 +44,16 @@ def initialize(armrest_configuration, options = {})
...
-          json_response = JSON.parse(response.body)
 
-          events = ArmrestCollection.new(
-            json_response['value'].map do |hash|
-              Azure::Armrest::Insights::Event.new(hash)
-            end
-          )
-
-          events.continuation_token = parse_skip_token(json_response)
+          klass  = Azure::Armrest::Insights::Event
```
Comment (bzwei) [inline lib/azure/armrest/insights/event_service.rb]: My thought is for `ArmrestCollection#initialize()` 
```
  def initialize(array, header = nil)
    # parse and process header
    super(array)
  end
```
then you only needs to append like this
```
events = ArmrestCollection.new(
  json_response['value'].map { |hash| Azure::Armrest::Insights::Event.new(hash) },
  response.header
)
```

Comment (djberg96) [inline lib/azure/armrest/insights/event_service.rb]: That doesn't deal with skip tokens (which are in the body), and doesn't provide a lot of added value IMO. My approach handles skip tokens and reduces a lot of code.

File: lib/azure/armrest/model/base_model.rb (hunk 2 above)
Comment (djberg96) [inline lib/azure/armrest/model/base_model.rb]: The response_headers need to be public IMO so that people can inspect rate limit information.

File: lib/azure/armrest/resource_group_based_service.rb [hunk 6]
```diff
@@ -35,19 +61,30 @@ def list(rgroup = configuration.resource_group)
...
         url = build_url
         url = yield(url) || url if block_given?
+
         response = rest_get(url)
-        results = JSON.parse(response)['value'].map { |hash| model_class.new(hash) }
+        results  = Azure::Armrest::ArmrestCollection.new(response, model_class)
+
         filter.empty? ? results : results.select { |obj| filter.all? { |k, v| obj.public_send(k) == v } }
+
+        results
```
Comment (bzwei) [inline lib/azure/armrest/resource_group_based_service.rb]: remove line 69 and 70

File: lib/azure/armrest/resource_group_based_service.rb [hunk 7]
```diff
@@ -104,16 +141,22 @@ def model_class
...
           response = rest_get(build_url(rg.name))
-          results = JSON.parse(response)['value'].map { |hash| model_class.new(hash) }
+          json_response = JSON.parse(response.body)['value']
+          headers = Azure::Armrest::ResponseHeaders.new(response.headers)
+          results = json_response.map { |hash| model_class.new(hash) }
           mutex.synchronize { array << results } unless results.blank?
         end
 
-        array.flatten
+        array = ArmrestCollection.new(array.flatten)
```
Comment (bzwei) [inline lib/azure/armrest/resource_group_based_service.rb]: will this work with the new ArmrestCollection initializer?

File: spec/insights_event_service_spec.rb [hunk 8]
```diff
@@ -6,7 +6,12 @@
 require 'spec_helper'
 
 describe "Insights::EventService" do
-  before { setup_params }
+  before do
+    setup_params
+    hash = {:x_ms_ratelimit_remaining_subscription_reads => '14999'}
+    allow_any_instance_of(String).to receive(:headers).and_return(hash)
```
Comment (bzwei) [inline spec/insights_event_service_spec.rb]: `allow_any_instance_of` needs to be avoided.

File: lib/azure/armrest/resource_group_based_service.rb (hunk 6 above)
Comment (djberg96) [inline lib/azure/armrest/resource_group_based_service.rb]: Done.

File: lib/azure/armrest/resource_group_based_service.rb (hunk 7 above)
Comment (djberg96) [inline lib/azure/armrest/resource_group_based_service.rb]: Nope, good catch. That was something I was tinkering with and forgot to get back to.

File: .gitignore [hunk 9]
```diff
@@ -7,3 +7,4 @@
 /pkg/
 /spec/reports/
 /tmp/
+/futz/
```
Comment (bzwei) [inline .gitignore]: where does this come from?

File: lib/azure/armrest/armrest_collection.rb [hunk 10]
```diff
@@ -4,6 +4,32 @@ module Azure
...
+        array.response_headers = response.headers
+        array.continuation_token = parse_skip_token(json_response)
+
+        array
+      end
+
+      alias skip_token continuation_token
+      alias skip_token= continuation_token=
+
+      private
```
Comment (bzwei) [inline lib/azure/armrest/armrest_collection.rb]: @djberg96 please fix. `private` does not make class method private. See the rubocop messages. Those should be addressed.

File: lib/azure/armrest/insights/event_service.rb [hunk 11]
```diff
@@ -44,22 +44,16 @@ def initialize(armrest_configuration, options = {})
...
+          events = Azure::Armrest::ArmrestCollection.create_from_response(response, klass)
```
Comment (bzwei) [inline lib/azure/armrest/insights/event_service.rb]: can line 55 and 56 combine into one line?

File: lib/azure/armrest/model/storage_account.rb [hunk 12]
```diff
@@ -99,11 +99,12 @@ def table_data(name, key = nil, options = {})
         key ||= properties.key1
 
         query = build_query(options)
-
         response = table_response(key, query, name)
-        json_response = JSON.parse(response.body)
 
-        data = ArmrestCollection.new(json_response['value'].map { |t| TableData.new(t) })
+        klass = Azure::Armrest::StorageAccount::TableData
+        data  = Azure::Armrest::ArmrestCollection.create_from_response(response, klass)
```
Comment (bzwei) [inline lib/azure/armrest/model/storage_account.rb]: same here, better one line?

File: lib/azure/armrest/resource_group_based_service.rb [hunk 13]
```diff
@@ -2,25 +2,51 @@ module Azure
...
-        model_class.new(response) unless response.empty?
+
+        obj = nil
+
+        unless response.empty?
+          obj = model_class.new(response.body)
+          obj.response_headers = Azure::Armrest::ResponseHeaders.new(response.headers)
+        end
+
+        obj
```
Comment (bzwei) [inline lib/azure/armrest/resource_group_based_service.rb]: Yours is fine. Does this look better?
```
model_class.new(response).tap |obj|
  obj.response_headers = Azure::Armrest::ResponseHeaders.new(response.headers)
end unless response.empty?
```

Comment (djberg96) [inline lib/azure/armrest/resource_group_based_service.rb]: That's harder for me to read.

File: lib/azure/armrest/insights/event_service.rb (hunk 11 above)
Comment (djberg96) [inline lib/azure/armrest/insights/event_service.rb]: I'm just trying to avoid extremely long lines.

File: lib/azure/armrest/model/storage_account.rb (hunk 12 above)
Comment (djberg96) [inline lib/azure/armrest/model/storage_account.rb]: I'm just trying to avoid extremely long lines.

File: .gitignore (hunk 9 above)
Comment (djberg96) [inline .gitignore]: Oops, I thought that would be ignored. Eh, it's harmless. It's just a place where I put local test scripts.

File: lib/azure/armrest/armrest_collection.rb (hunk 10 above)
Comment (djberg96) [inline lib/azure/armrest/armrest_collection.rb]: @bzwei Fixed. It does if I use a class/self block. :)
